### PR TITLE
Adding the option for specify the project title on packaged project

### PR DESCRIPTION
**Pull request by SeqLaz**
This adds the option to specify a new project name and title in the exported project when packaging.

| Before | After |
|--------|-------|
| Before Image | After Image |

Related PR: https://github.com/opengisch/libqfieldsync/pull/94

### CODE REVIEW

File: qfieldsync/ui/package_dialog.ui
Comment (suricactus) [inline qfieldsync/ui/package_dialog.ui:25]: This should not be HTML.

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:196]: I guess you don't need it in this PR, do you?

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:76]:
```suggestion
        self.original_project_title = project.title()
```

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:77]:
```suggestion
        self.packaged_project_title.setText(self.original_title_project)
```

File: qfieldsync/gui/package_dialog.py
Comment (SeqLaz) [reply qfieldsync/gui/package_dialog.py:196]: Due this was removed on the other [PR](https://github.com/opengisch/qfieldsync/pull/616) now this will be necessary there

File: qfieldsync/ui/package_dialog.ui
Comment (SeqLaz) [reply qfieldsync/ui/package_dialog.ui]: Good catch!

File: qfieldsync/ui/package_dialog.ui
Comment (suricactus) [inline qfieldsync/ui/package_dialog.ui:27]:
```suggestion
```

I guess the HTML problem above comes from the fact that the QLabel is `RichText`.

File: qfieldsync/ui/package_dialog.ui
Comment (suricactus) [inline qfieldsync/ui/package_dialog.ui:49]: Same as above.

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:78]:
```suggestion
        self.packaged_project_title.setText(project.title())
```

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:73]: This is only for internal use, right? Can we make it a "private" property by prefixing with underscore?
```suggestion
        self._original_project_path = self.project.fileName()
```

Also, please move it next to the rest of the changes in this PR a few lines below.

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:79]: What if the project we are trying to package is not saved yet? Imagine you open a new blank project, set all the layers, config etc and then press "package project" without ever saving it to the disk?

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:80]:
```suggestion
        self.tmp_project_filename = make_temp_qgis_file(self.project)
```

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:176]:
```suggestion
        tmp_project_filepath = Path(self.tmp_project_filename).parent
```

Or whatever is the syntax using `pathlib`. Note that in all the codebase we prefer using `pathlib` over `os.path`.

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:205]: I think this is already handled internally in the `OfflineConverter`. It should reopen the original project internally, no? If yes, then we need to make sure the OfflineConverter opens the right project.

**Review by suricactus — COMMENTED** [review]
Instead of complicating the code in the caller (in this case QFieldSync), we should add this functionality within the OfflineConverter itself. Why? Because the converter already does the copy of the file.

For example here we have hardcoded the name of the output file: https://github.com/opengisch/libqfieldsync/blob/master/libqfieldsync/offline_converter.py#L276

What we can do is to pass the output name to be a parameter when we construct the `OfflineConverter` class. We can even replace the `export_dir` param with `export_filename` param, and then derive the `export_dir` to be the path to `export_filename`.

**Comment by suricactus** [comment]
Please note that most of the suggestion here need code-wide changes, the suggestions are only pointing to the highlights.

File: qfieldsync/gui/package_dialog.py
Comment (SeqLaz) [reply qfieldsync/gui/package_dialog.py]: This is done in this [PR](https://github.com/opengisch/libqfieldsync/pull/94)

**Comment by SeqLaz** [comment]
@suricactus The changes and suggestions were addressed in this [PR](https://github.com/opengisch/libqfieldsync/pull/94)

**Comment by SeqLaz** [comment]
> Please note that most of the suggestion here need code-wide changes, the suggestions are only pointing to the highlights.

@suricactus The changes and suggestions were addressed in this [PR](https://github.com/opengisch/libqfieldsync/pull/94)

**Comment by suricactus** [comment]
Few comments after the libqfieldsync review:
- Drop the "export directory" and export finame in favor of single "export filename" with the full path
- After clicking "Create" OR when changing the value of "export filename", make sure the user is notified there is already a file with that name in that location.
- Prepopulate the "export_filename" with the default value of "export directory" with "/" and current project basename and "_qfield" suffix and ".qgs" extension. You can create a utility function in libqfieldsync `get_export_filename_suggestion(project.fileName())`

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:188]:
```suggestion
        packaged_project_filename = Path(self.packaged_project_filename.text())
```

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:174]:
```suggestion
                self.tr("File Already Exists"),
                self.tr('File "{}" already exists. Do you want to overwrite it?').format(export_packaged_project.name),
```
don't forget the translation

File: qfieldsync/ui/package_dialog.ui
Comment (suricactus) [inline qfieldsync/ui/package_dialog.ui:88]:
```suggestion
           <widget class="QLineEdit" name="packagedProjectTitleLineEdit"/>
```
Note that the names here are camelCase for historical reasons. Don't mix multiple naming conventions.

File: qfieldsync/ui/package_dialog.ui
Comment (suricactus) [reply qfieldsync/ui/package_dialog.ui:110]:
```suggestion
           <widget class="QgsFileWidget" name="packagedProjectFileWidget"/>
```

File: qfieldsync/ui/package_dialog.ui
Comment (suricactus) [inline qfieldsync/ui/package_dialog.ui:108]: Why do we have the `manualDir` here? Also I think there is some kind of an error with the XML, or I miss somehow the closing `</item>` for line 93. How did you edit this?

**Review by suricactus — CHANGES_REQUESTED** [review]
Not there yet.

**Comment by SeqLaz** [comment]
> Not there yet.

Hey @suricactus please have a look.
I squashed due to the PR had a lot of merge. In the last commit I addressed all the suggestion and some improvements.

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:149]:
```suggestion
    def get_export_filename_suggestion(self) -> str:
```

Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:149]:
```suggestion
    def get_export_filename_suggestion(self) -> str:
```

Please add typing on all new methods you add.

**Comment by SeqLaz** [comment]
Hey @suricactus I added the typing hints on the new functions I added, have a look!

File: qfieldsync/gui/package_dialog.py
Comment (suricactus) [inline qfieldsync/gui/package_dialog.py:159]: I failed to find who uses this? In any case, such functions should not exist, it's better to do it explicitly what you do in the function body.

Comment (SeqLaz) [reply qfieldsync/gui/package_dialog.py]: You are completely right, the function is not needed!

**Comment by SeqLaz** [comment]
Hey @suricactus, I deleted the unnecessary function and joined the functions of export_folder and get_filename_suggestion.

**Review by suricactus — APPROVED** [review]
Great work!

**Review by suricactus — CHANGES_REQUESTED** [review]
Ah, just realized, you never bumped libqfieldsync here. Last step before merging. I guess the tests will fail here too.

**Comment by suricactus** [comment]
@SeqLaz you have a conflict now. I will resolve it to unblock the tests.

**Comment by SeqLaz** [comment]
Hey @suricactus I've fixed the tests!.

**Pull request merged**
